### DISCUSSION TITLE
Odd behaviour of on-demand revalidation

### DISCUSSION THREAD

**Discussion by Normaly0** [vercel/next.js#52318] [Help] [answered]
### Summary

Hey and thanks for taking your time and reading this!

I'm currently finishing up work on a bigger NextJS project and wanted to implement on demand revalidation to improve the life of our editors.

The NextJS project is hosted on the app platform of DigitalOccean.
And I'm using a Strapi CMS as the backend.

The following is my revalidate.js:

```
export default async function handler(req, res) {

    // Check for secret to confirm this is a valid request
    if (req.headers.authorization !== process.env.REVALIDATION_SECRET) {
        return res.status(401).json({ message: 'Invalid authorization header' })
    }

    try {

        switch(req.body.model) {
            case 'agency-page':
                res.revalidate('/agency');
                break;
            case 'client-page':
                res.revalidate('/clients');
                break;
            case 'contact-page':
                res.revalidate('/contact');
                break;
            case 'data-privacy':
                res.revalidate('/data-privacy');
                break;
            case 'home-page':
                res.revalidate('/');
                break;
            case 'legal-notice':
                res.revalidate('/legal-notice');
                break;
            case 'method-page':
                res.revalidate('/method');
                break;
            case 'partner-page':
                res.revalidate('/partners');
                break;
            case 'service-page':
                res.revalidate('/services');
                break;
            case 'work-page':
                res.revalidate('/work');
                break;
            default: 
                console.log(req.body.model);
                break;
        }

        console.log('REVALIDING PAGE: ' + req.body.model)
        
        return res.json({ revalidated: true })

    } catch (err) {

        // If there was an error, Next.js will continue
        // to show the last successfully generated page
        return res.status(500).send('Error revalidating')

    }
}
```

The entire revalidation process is triggered on any change of content inside Strapi with their build in `webhooks`.

So when testing all of this locally it works like a charm! 
Whenever I update any content that change pretty much instantly propagates to the NextJS application.

But it's a different and very....odd story on the staging website.
Strapi's side works without issues, since I am logging the module that should be updated in my code, I can confirm that the `revalidate.js` code is running correctly and the function is being called.

However those changes don't take effect fully. 
When I refresh the page I still see the old version.
But if I use the `Link` navigation within the application I am being shown the new version....
On top of it all it only works for one single update, after the page has been updated a second time not even the `Link` navigation gives me the new version of the page.

I'm kinda....very confused so any kind of ideas or feedback is very much appreciated!

**Comment by icyJoseph** [comment]
Have you tried `await`ing the `res.revalidate` call?

Also, have you gotten a CDN or Service worker that might be caching data?

**Reply by Normaly0** [reply]
Thanks a lot for pointing that out!
Changing it to an `await` had some....ironic results.

Now the situation is flipped!
The first reload gives me the updated data.
Link navigation shows the stale data

But regardless of that again only the very first revalidation actually changes content.
It doesn't matter how often I update the data afterwards, despite triggering a revalidation the old content is still shown.

Looking at the headers of both request they both have `Cache-Control: s-maxage=31536000, stale-while-revalidate`

From my knowledge the DigitalOcean App platform doesn't have any CDN's attached to it nor can I change the server configuration to begin with..

It's just...weird that the NextJS `Link` gives me a different result to a full page reload too.

**Reply by icyJoseph** [reply]
Yeah, well, this is where the Next.js magic happens, I know in the midst of a problem is hard to see it as magic...

You see, even though Next.js does pre-render pages aggressively, once you are on the app client side, navigations through next/links, are essentially CSR.

In order to do the CSR swap, Next.js downloads the JS for the page, and a JSON file which are the initial props for that page. This is called prefetching. For pages that used GetServerSideProps, only the JS is prefetched, and the JSON is downloaded on click.

We can infer, that the pre-rendering step generates an HTML page, and JS+JSON payload combo to drive the page on CSR. 

So there's actually two version of your page, and, normally, the JS+JSON combo, when executed, outputs the same HTML page. Otherwise, we run into hydration errors. 

A precision would be that the HTML page has a script tag where the initial JSON data is copied to. So that hydration on first page render, can happen without waiting for an extra JSON request.

So, if the revalidate only generated a new HTML page and not a new JSON payload, it'd make sense that you get different results from the Link and full page reload.

As to why the content stops changing, I don't know... Maybe the Next.js process lacks some permissions or something like that? Guessing at this point.

**Reply by Normaly0** [reply]
Huh, thanks for the in-depth explanation of it all.
I was lowkey aware of this concept altho not of the fact that it essentially turns into CSR.

Now after testing some more I can say that the results of which is the newer content are completely random!
Some pages have the updated content on their `Link` navigations, while others have it on refresh.....

And again
It only ever updates one single time anyway

I'll do some more tests tomorrow but I believe when I still had the `revalidate: 30` property set on my `getStaticProps` everything was working fine.

So worst case I'll have to move back to that, altho solving this issue would save me some sleepless nights.

**Comment by Normaly0** [comment]
As a lil follow up on this.
Since I wanted to see if this is only an issue on DigitalOcean I deployed my project on Vercel.

And sure enough there are no issues!
Comparing the `Cache-Control` headers of the two json request there is a difference from DO to Vercel.

DigitalOcean:
<img width="381" alt="Screenshot 2023-07-07 at 14 14 00" src="https://github.com/vercel/next.js/assets/40768391/a0c970e5-cc21-4110-ad4e-ebc117c0a5ef">


Vercel:
<img width="424" alt="Screenshot 2023-07-07 at 14 12 54" src="https://github.com/vercel/next.js/assets/40768391/b8542d9f-5984-49c4-9aa6-ebd1bad8845c">

So it seems that this might be related to the issue?
However checking the NextJS documentation it states that
`if revalidate is not used, s-maxage=31536000, stale-while-revalidate to cache for the maximum age possible`

Meaning that NextJS is correctly setting the cache header, beats me why it's different on Vercel and why the manual revalidation is actually working.

**Reply by icyJoseph** [reply]
Aha, so my CDN hunch was somewhat correct:

- https://vercel.com/docs/concepts/edge-network/caching#cdn-cache-control
> If you set Cache-Control without a CDN-Cache-Control, the Vercel Edge Network strips s-maxage and stale-while-revalidate from the response before sending it to the browser. To determine if the response was served from the cache, check the [x-vercel-cache](https://vercel.com/docs/concepts/edge-network/caching#x-vercel-cache) header in the response.

**Reply by Normaly0** [reply, marked as answer]
Very much so!
Now I've done some more digging and found the following post talking about a CDN loop caused by the `Cache-Headers`
https://github.com/vercel/next.js/discussions/35104#discussioncomment-5801921

Very glad that they mentioned DigitalOcean in it, altho sadly the only way to solve this issue for them was to move to Vercel.

But I might have found a....accidental solution.
I wanted to test if the same happens if I set `revalidation: 10` and in doing so forgot to disable the webhook on my backend.
Wouldn't you know it....it works now!

And I mean
The content updates _before_ the 10 seconds timer is actually up! So for some reason the `revalidation()` function now works correctly.

I've tested some more and it seems that the highest I can set the revalidation timer to is 15 seconds. After that I start to get the exact same issue again until the application updates through the rebuild triggered by the timer.

So...certainly a weird one but this might be the only solution without moving to Vercel.
I'll monitor the server load to see if the requests every 15 seconds are too much for it.
